metagraph/src/graph/representation/succinct/boss_chunk_construct.cpp

```diff
@@ -669,25 +711,36 @@ BOSS::Chunk construct_boss_chunk_disk(KmerCollector &kmer_collector,
 
     size_t k = kmer_collector.get_k() - 1;
     const std::filesystem::path dir = kmer_collector.tmp_dir();
-    size_t num_threads = kmer_collector.num_threads();
+    size_t num_threads = std::max(kmer_collector.num_threads(), (size_t)1);
     const uint64_t buffer_size = kmer_collector.buffer_size();
 
     auto &container = kmer_collector.container();
     const std::vector<std::string> chunk_fnames = container.get_chunks();
 
+    const size_t A2 = std::pow(KmerExtractor2Bit().alphabet.size(), 2);
+    size_t n_threads = check_fd_and_adjust_threads(std::min(num_threads, chunk_fnames.size()),
+                                                   (A2 + 1) * (2 + utils::is_pair_v<T>));
     // split each chunk by F and W
     std::vector<std::vector<std::string>> chunks_split(chunk_fnames.size());
-    #pragma omp parallel for num_threads(num_threads) schedule(dynamic)
+    #pragma omp parallel for num_threads(n_threads) schedule(dynamic)
     for (size_t i = 0; i < chunk_fnames.size(); ++i) {
         chunks_split[i] = split<T_REAL>(k, chunk_fnames[i]);
     }
 
     // for a DNA alphabet, this will contain 16 chunks, split by kmer[0] and kmer[1]
-    std::vector<std::string> real_F_W(std::pow(KmerExtractor2Bit().alphabet.size(), 2));
+    std::vector<std::string> real_F_W(A2);
+
+    n_threads = check_fd_and_adjust_threads(std::min(num_threads, real_F_W.size()),
+                                            4 * (2 + utils::is_pair_v<T>));
```

Comment:
is the `4` here supposed to be the alphabet size?

Reply:
no, it's 2 + 2 (2 chunks with 2 files per each)